### ISSUE TITLE
[naming-convention] Allow underscore prefix for unused parameters

### ISSUE THREAD

**Issue by glen-84** [typescript-eslint/typescript-eslint#1510]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": "default",
        "format": ["strictCamelCase"]
      }
    ]
  }
}
```

```TS
onComplete: (_id, name, responseJson) => {
    // Using name and responseJson, but not id.
}
```

**Expected Result**

No error on `_id`.

**Actual Result**

Error on `_id`.

**Additional Info**

As mentioned [here](https://github.com/typescript-eslint/typescript-eslint/pull/1318#issuecomment-564015707) with Brad's response [here](https://github.com/typescript-eslint/typescript-eslint/pull/1318#issuecomment-564300069).

When using the `noUnusedParameters` compiler option, you have to use an underscore to suppress TypeScript's error. However, this shouldn't be done unless the parameter _is_ actually unused.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.16.0` |
| `@typescript-eslint/parser`        | `2.16.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.14.1` |
| `npm`                              | `6.13.4` |

**Comment by bradzacher** [comment]
I need to figure out how to do this in a performant way. Using typescript's diagnostics would be awesome, but they're super slow - ~50ms to get the diagnostics for a file, which is too slow for linting (related: #1335).

I should test tsutils's unused variables function. If that's quick, then that's a decent compromise.

The only problem with both of these routes is that without #716, this would require type information. Though we're planning on starting the 3.0.0 release soon, so that isn't that far off.

**Comment by nstepien** [comment]
It would also be nice to allow `_` as a valid parameter name:
```js
Array(length).fill(0).map((_, i) => i);
```

**Comment by glen-84** [comment]
@nstepien,

You can allow that by setting `leadingUnderscore` to `allow` for the `parameter` selector:

```js
{
    selector: "parameter",
    format: ["strictCamelCase"], // Change if you want to.
    leadingUnderscore: "allow"
}
```

**Comment by bradzacher** [comment]
@nstepien this is already allowed by just setting `leadingUnderscore: 'allow'`.

[From the docs on format options:](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md#format-options)

> One final note is that if the name were to become empty via this trimming process, it is considered to match all formats.

**Comment by moulikcipherX** [comment]
Hy @bradzacher  I think there is nothing to work in this issue so we can close it.

**Comment by bradzacher** [comment]
Sorry, the last few comments were unrelated to the root issue.

For a bit of background; the `naming-convention` rule was derived from a [3rd party rule made for tslint](https://github.com/ajafff/tslint-consistent-codestyle/blob/master/docs/naming-convention.md).

When I wrote it, I cut some features from that rule for the purpose of keeping complexity down and cutting features until we saw how people used them.
One feature I cut was the `unused` modifier. This modifier would cause a selector to only match unused variables.

Right now you can only say "all variables can have leading underscores".
With this modifier, it would allow you to instead say "only unused variables can have leading underscores".

**Comment by moulikcipherX** [comment]
So, I can work on it.

**Comment by bradzacher** [comment]
This needs some exploration to find an efficient way to check if a variable is unused. You're certainly welcome to explore this if you'd like.

I don't have a clear idea of how to do this in a performant manner yet.

**Comment by fabien-h** [comment]
Have you found an efficient way of doing it ?

```
    {
        "selector": "property",
        "format": ["camelCase"],
        "leadingUnderscore": "allow",
        "trailingUnderscore": "forbid"
      },
```

Still raise an error.

**Comment by bradzacher** [comment]
Once #2039 lands this should be pretty simple to do.

**Comment by Thomazella** [comment]
@bradzacher nice, #2039 got merged :100:

**Comment by evanjmg** [comment]
How do we use it? I can't get this to work

**Comment by glen-84** [comment]
@evanjmg Use what? This hasn't been implemented yet.
